<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Pengetahuan-Dasar-Python" data-toc-modified-id="Pengetahuan-Dasar-Python-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Pengetahuan Dasar Python</a></span><ul class="toc-item"><li><span><a href="#Virtual-Environtment" data-toc-modified-id="Virtual-Environtment-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Virtual Environtment</a></span></li><li><span><a href="#Install-Kernel-pada-Virtual-Environtment" data-toc-modified-id="Install-Kernel-pada-Virtual-Environtment-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Install Kernel pada Virtual Environtment</a></span></li></ul></li><li><span><a href="#Struktur-Data-pada-Python" data-toc-modified-id="Struktur-Data-pada-Python-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Struktur Data pada Python</a></span><ul class="toc-item"><li><span><a href="#List" data-toc-modified-id="List-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>List</a></span></li><li><span><a href="#Tuple" data-toc-modified-id="Tuple-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Tuple</a></span></li><li><span><a href="#Dictionary" data-toc-modified-id="Dictionary-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Dictionary</a></span></li><li><span><a href="#Quiz-1" data-toc-modified-id="Quiz-1-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Quiz 1</a></span></li></ul></li><li><span><a href="#Module-pandas" data-toc-modified-id="Module-pandas-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Module <code>pandas</code></a></span><ul class="toc-item"><li><span><a href="#Data-Frame" data-toc-modified-id="Data-Frame-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Data Frame</a></span></li><li><span><a href="#Tipe-tipe-Data" data-toc-modified-id="Tipe-tipe-Data-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Tipe-tipe Data</a></span><ul class="toc-item"><li><span><a href="#Tipe-Data-Categorical-dan-Numerical" data-toc-modified-id="Tipe-Data-Categorical-dan-Numerical-3.2.1"><span class="toc-item-num">3.2.1&nbsp;&nbsp;</span>Tipe Data Categorical dan Numerical</a></span></li><li><span><a href="#Quiz-2" data-toc-modified-id="Quiz-2-3.2.2"><span class="toc-item-num">3.2.2&nbsp;&nbsp;</span>Quiz 2</a></span></li></ul></li><li><span><a href="#Eksplorasi-Data" data-toc-modified-id="Eksplorasi-Data-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Eksplorasi Data</a></span><ul class="toc-item"><li><span><a href="#Quiz-3" data-toc-modified-id="Quiz-3-3.3.1"><span class="toc-item-num">3.3.1&nbsp;&nbsp;</span>Quiz 3</a></span></li></ul></li><li><span><a href="#Indexing-dan-Subsetting" data-toc-modified-id="Indexing-dan-Subsetting-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Indexing dan Subsetting</a></span><ul class="toc-item"><li><span><a href="#Kondisional-Subset" data-toc-modified-id="Kondisional-Subset-3.4.1"><span class="toc-item-num">3.4.1&nbsp;&nbsp;</span>Kondisional Subset</a></span></li><li><span><a href="#Quiz-4" data-toc-modified-id="Quiz-4-3.4.2"><span class="toc-item-num">3.4.2&nbsp;&nbsp;</span>Quiz 4</a></span></li></ul></li></ul></li><li><span><a href="#Task" data-toc-modified-id="Task-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Task</a></span></li></ul></div>

**LearnPy: Python for Data Analyst**

* Author : Team Algoritma
* Developed by [Algoritma](https://algorit.ma)'s product division and instructors team

---

# Pengetahuan Dasar Python

Hi *future Data Analyst*!! Pada notebook ini kita akan membahas mengenai analisis data menggunakan bahasa pemrograman Python menggunakan salah satu module yang cukup powerfull yaitu `pandas`. Namun sebelum itu kita akan membahas beberapa hal terlebih dahulu sebelum masuk ke pengenalan kita terhadap `pandas`.

## Virtual Environtment

Virtual environtment merupakan alat yang digunakan untuk menyimpan dependensi dari suatu library atau module yang diperlukan ketika mengerjakan suatu project tertentu. Beberapa project tertentu membutuhkan dependency python yang berbeda, oleh karena itu cukup menyulitkan jika kita membuat virtual environtment. 

Beberapa hal yang perlu dipersiapkan dalam membuat virtual environtment yaitu:
- sudah memiliki Anaconda Navigator/Miniconda
- menggunakan CLI (command line interface) sebagai alat untuk membuat, memanage, dan mengaktifkan virtual environtment

Ketika kita sudah memiliki Anaconda Navigator, biasanya tersedia CLI bawaan dari Anaconda Navigator yaitu Anaconda Prompt. Kita akan coba membuat suatu virtual environtment menggunakan Anaconda Prompt. Setelah dibuka Anaconda Prompt, biasanya langsung muncul tampilan sebagai berikut.

![](img/base.PNG)

Saat pertama kali kita membuka Anaconda Prompt, kita akan diberikan informasi bahwa kita masih didalam environtment `base` atau environtment secara general. Untuk bisa membuat virtual environtment, kita bisa menggunakan code berikut.
    
    
    conda create --name mynewenv python=3.7
    

![](img/create-new.PNG)

Tujuan dari kita menuliskan `python=3.7` agar python yang terinstall pada virtual environtment yang kita buat adalah python dengan versi 3.7. Inilah salah satu keuntungan yang kita peroleh dengan membuat virtual environtment, kita dapat mengatur versi python yang kita inginkan sesuai kebutuhan dari project yang ingin kita buat. 
Selanjutnya akan muncul beberapa dependency library yang minta untuk di install saat membuat virtual environtment baru. Kita bisa mengetikkan huruf `y` untuk melanjutkan proses pembuatan virtual environtment yang baru.

![](img/dependency.PNG)

Setelah virtual environtment yang baru kita buat sudah selesai, kita bisa mengaktifkan environtment yang sudah kita buat dengan code berikut.

```dash
conda activate mynewenv
```
Kita dapat melihat beberapa module yang sudah terinstall pada virtual environtment yang kita buat dengan menggunakan code berikut.

```dash
conda list
```

![](img/conda-list.PNG)

## Install Kernel pada Virtual Environtment

Kernel merupakan suatu penghubung antara environtment dengan setiap project yang kita gunakan. Jika dimisalkan sebuah rumah yang memiliki ruangan, kernel merupakan pintu penghubung yang bisa digunakan untuk bisa mengakses ruangan tersebut. Secara default apabila kita tidak melakukan install kernel pada environtment yang kita buat maka kernelnya adalah `Python 3`. Namun apabila kita menggunakan kernel yang sudah kita buat untuk masing-masing environtment yang kita miliki maka nama kernelnya akan menyerupai nama environtment yang sudah kita buat. 

Kita dapat melakukan install kernel untuk environtment yang sudah kita buat yaitu `mynewenv` dengan cara sebagai berikut.

- Pastikan sudah memiliki `ipykernel` pada environtment terlebih dahulu, apabila belum bisa lakukan installasi `ipykernel` dengan cara berikut:

``` dash
conda install ipykernel # menggunakan conda
pip install ipykernel # menggunakan pip
```

- Setelah terinstall, maka kita lakukan pengaktifan `ipykernel` pada environtment kita yaitu `mynewenv` :

```dash
python -m ipykernel install --user --name mynewenv`
```

Setelah sukses melakukan penginstallan akan ada status pada CLI sebagaimana gambar dibawah ini dengan begitu maka install kernel untuk environtment yang kita gunakan sudah selesai. 

![](img/ipykernel-installed.png)

Kita bisa melanjutkan dengan membuka python kita menggunakan IDE yang kita miliki yaitu bisa melalui Jupyter Notebook, Jupyter Lab, atau IDE yang lain. Jika ingin menggunakan IDE Jupyter Notebook, maka cukup gunakan code `jupyter notebook` dan apabila menggunakan IDE Jupyter Lab gunakan code `jupyter lab` pada CLI. 

Sebelum mengenal lebih jauh beberapa keguanaan yang bisa kita dapatkan dari module `pandas`, kita akan coba mengenal beberapa object di python.

# Struktur Data pada Python

Dalam python sendiri terdapat sebuah struktur data, dimana struktur data ini biasanya disusun berdasarkan satu dimensi atau dua dimensi. Terdapat beberapa struktur data pada python diantaranya yaitu :
- list
- tuple
- dictionary

## List

List adalah suatu struktur data pada python yang tersusun secara satu dimensi dimana nilai dari dari anggotanya bisa kita ganti. List secara umum bisa disebut sebagai "array" pada python. List dapat dibuat dengan menggunakan tanda bracket atau `[..,..,..]`. Perhatikan list dibawah ini.

In [1]:
list_angka = [1,2,3,4,5,6,7,8,9]
print(list_angka)

[1, 2, 3, 4, 5, 6, 7, 8, 9]


List bisa kita ganti nilai pada anggotanya dengan melakukan inspek urutan pada struktur list. Misalnya kita ingin mecoba mengganti nilai 6 pada object `list_angka` diatas menjadi 10. Kita harus tahu terlebih dahulu nilai 6 pada `list_angka` terdapat pada urutan ke berapa. Pada python, index suatu data dimulai dari angka 0, sehingga kita bisa melakukan inspek urutan data dengan cara $index = data_i-1$. Selanjutnya untuk melakukan inspek salah satu nilai pada list dapat menggunakan `nama_object[index]`.

In [2]:
list_angka[5]

6

In [3]:
list_angka[5]=10

In [4]:
print(list_angka)

[1, 2, 3, 4, 5, 10, 7, 8, 9]


## Tuple

Tuple adalah suatu struktur data pada python yang tersusun satu dimensi hanya saja nilai dari anggotanya konstan dan tidak dapat diubah. Untuk dapat membuat suatu tuple bisa menggunakan tanda kurung atau `(..,..,..)` atau tanpa tanda kurung. Perhatikan contoh tuple dibawah ini.

In [5]:
tuple_object = ('Learn', 'Python', 2020) 
tuple_object

('Learn', 'Python', 2020)

In [6]:
tuple_object2 = 'list', 'tuple', 'dictionary'
print(tuple_object2)

('list', 'tuple', 'dictionary')


Tuple biasanya lebih mudah dikenal sebagai object string, dimana hanya berisi sederet informasi yang memiliki satu dimensi. Untuk mengakses index tertentu pada tuple dapat menggunakan `nama_object[]` seperti pada list dan apabila kita ingin mengetahui suatu nilai pada tuple terdapat pada index ke berapa, maka kita bisa gunakan `nama_object.index()`.

In [7]:
tuple_object[2]

2020

In [8]:
tuple_object2.index('tuple')

1

Selain itu, jika kita memiliki lebih dari 1 nilai pada object tuple yang kita miliki, kita bisa melakukan slicing pada object tuple kita dengan menggunakan `tuple[]` dan memasukkan urutan nilai yang ingin dimasukkan.

**Ingat!!** Pada python untuk urutan angka dimulai dari angka 0.

In [9]:
print(tuple_object[0])
print(tuple_object2[1:3])

Learn
('tuple', 'dictionary')


Nilai dari tuple juga tidak bisa kita ubah semau kita. Apabila kita mengubah nilainya, maka akan ada error yang dimunculkan dari python.

In [10]:
tuple_object[1] = 100
print(tuple_object[1])



TypeError: 'tuple' object does not support item assignment

## Dictionary

Dictionary adalah suatu struktur data pada python yang terdiri dari *key* dan *value* dimana *key* bisa kita tentukan sendiri beserta *value* bisa kita ubah nilainya. *Key* pada dictionary memiliki maksud suatu index sedangkan *value* adalah nilai dari index pada struktur data dictionary. Setiap satu *key* pasti memiliki satu *value*. Dalam pembuatan dictionary kita bisa menggunakan tanda kurung kurawal atau `{key:value}`. Perhatikan dictionary dibawah ini.

In [12]:
dict_object = {1:'merah', 2:8, 3:'hijau', 4:10, 5:'hitam', 'apel':9}
dict_object

{1: 'merah', 2: 8, 3: 'hijau', 4: 10, 5: 'hitam', 'apel': 9}

Untuk bisa melakukan inspek elemen pada dictionary kita bisa gunakan `nama_object[key]` seperti pada list dan tuple. 

In [13]:
print(dict_object[5])
print(dict_object['apel'])

hitam
9


Untuk bisa mengubah nilai atau *value* kita bisa mengakses berdasarkan *key* yang ingin kita ubah. Misalnya disini kita coba ubah *value* pada *key* apel yang tadinya 9 menjadi kata "fuji".

In [14]:
dict_object['apel'] = 'fuji'
print(dict_object)

{1: 'merah', 2: 8, 3: 'hijau', 4: 10, 5: 'hitam', 'apel': 'fuji'}


## Quiz 1

1. Berdasarkan pilihan dibawah ini, manakah struktur data yang bisa kita tentukan index dan nilainya?
    - [ ] list
    - [ ] tuple
    - [v] dictionary
    
#### Dictionary adalah suatu struktur data pada python yang terdiri dari key dan value dimana key bisa kita "tentukan" sendiri beserta value bisa kita ubah nilainya

---

2. Jika ada struktur data dibawah ini.

```py
object1 = [1,45,23,12,45,2,3,4,67,23]
```
Termasuk apakah `object1` diatas?

- [v] list
- [ ] tuple
- [ ] dictionary

#### list akan diawali dengan tanda [kurung siku]

---

3. Manakah dari pilihan dibawah ini yang tidak menghasilkan error?
    - [ ] `object1 = (1,2,3,4,5,6)
      object1[3] = 9
      print(object1)`
    - [ ] `object1 = {'angka':1, 'bulan':03, 'tahun': 2020, 'lokasi': 'jakarta'}
      object1 = {'angka':1, 'angka':4, 'angka':10, 'bulan':03, 'tahun': 2020, 'tahun1': 2019, 'tahun': 2021, 'lokasi': 'jakarta'}`
    - [v] `object1 = [1,2,3,4,5,6]
      object1.index(3)`
    - [ ] `object1 = [1,2,3,4,5,6]
      object1.index(0)`
      
#### 𝑖𝑛𝑑𝑒𝑥 = 𝑑𝑎𝑡𝑎𝑖 − 1
     
---

4. Berapakah nilai yang dihasilkan dari code dibawah ini?

```py
angka = {1:10, 2:20, 3:30, 4:40, 5:50, 6:60, 6:90, 7:70, 8:80, 8:88, 9:90}
print(angka[1], angka[6], angka[8])
```

- [ ] 10, 50, 88
- [ ] 10, 60, 88
- [v] 10, 90, 88
- [ ] 10, 90, 80

#### Setiap satu key pasti memiliki satu value, seperti yang sudah-sudah statement terakhir yang dipakai. Jadi Key nomor 6 dan 8 menggunakan argumen terakhir nilainya, sementara argumen yang awal tidak lagi dianggap indeks ataupun key

# Module `pandas`

Selanjutnya kita akan coba mengeksplorasi module `pandas` untuk keperluan analisis data. Untuk dapat menggunakan module tersebut, kita harus melakukan import module terlebih dahulu. Untuk melakukan import module cukup menggunakan code `import`.

In [15]:
import pandas as pd
print(pd.__version__)

1.0.1


## Data Frame 

Kita akan coba untuk melakukan exploratory data menggunakan data turnover suatu perusahaan. Untuk melihat datanya kita harus mengimport data turnover tersebut yang tersimpan dengan nama file `turnover_balance.csv`. Data ini menunjukkan suatu record history karyawan pada perusahaan tertentu yang mengalami turnover berdasarkan beberapa atribut-atribut penunjangnya yaitu `satisfaction_level`, `last_evaluation`, `number_project`, `average_monthly_hours`, `time_spend_company`, `Work_accident`, `promotion_last_5years`, `division`, dan `salary`. Berikut data turnover tersebut :

In [16]:
turnover = pd.read_csv(r"C:\Users\Lenovo\Desktop\Machine Learning\data\turnover_balance.csv")
turnover.head()

## Modif sikit folder penyimpanan datanya

satisfaction_level  last_evaluation  number_project  average_montly_hours  \
0                0.82             0.68               3                   140   
1                0.79             0.67               5                   156   
2                0.73             0.95               3                   149   
3                0.92             0.78               3                   218   
4                0.69             1.00               5                   237   

   time_spend_company  Work_accident  promotion_last_5years     division  \
0                   2              0                      0        sales   
1                   2              0                      0  product_mng   
2                   2              0                      0      support   
3                   3              0                      0    technical   
4                   3              0                      0    technical   

  salary  left  
0    low     0  
1    low     0  
2    low     0  
3    low     0  
4   high     0

Berdasarkan data `turnover` diatas, berikut ini adalah glimpse dari setiap variabelnya :

* `satisfaction_level` : tingkat kepuasan karyawan bekerja di suatu perusahaan
* `last_evaluation` : tingkat kepuasan karyawan pada evaluasi terakhir
* `number_project` : banyaknya project yang telah diterima karyawan
* `average_monthly_hours` : rata-rata jam kerja per bulan
* `time_spend_company` : lama waktu bekerja dalam perusahaan (tahun)
* `Work_accident` : ada tidaknya kecelakaan kerja, 0 = tidak ada, 1 = ada
* `promotion_last_5years` : pernah mendapatkan promosi dalam 5 tahun terakhir, 0 = tidak, 1 = ya
* `division` : nama departemen atau divisi
* `salary` : tingkat pendapatan, dibagi menjadi low, medium dan high
* `left` : data history karyawan resign, 0 = tidak, 1 = ya

Sebelum melangkah pada hal-hal yang rumit, kita akan coba untuk mengulik dari hal yang sederhana terlebih dahulu.

## Tipe-tipe Data 

Sebagai seorang data analyst, kita harus mengetahui terlebih dahulu tipe-tipe data yang akan kita analisis. Pada python, ketika kita bekerja menggunakan module `pandas` untuk melakukan pengecekan tipe data pada data yang digunakan, kita bisa menggunakan `dtypes`. Nilai yang akan dikeluarkan oleh `dtypes` adalah sebuah series yang berisi tipe data setiap variabel pada suatu data frame.

In [17]:
turnover.dtypes

satisfaction_level       float64
last_evaluation          float64
number_project             int64
average_montly_hours       int64
time_spend_company         int64
Work_accident              int64
promotion_last_5years      int64
division                  object
salary                    object
left                       int64
dtype: object

Tipe-tipe data yang terdapat pada Python umumnya adalah:

* **float** : tipe data yang berisi angka kontinu
* **int** : tipe data yang berisi angka bulat
* **object** : tipe data yang berisi alphabetic, seperti nama, alamat, dst
* **bool** : tipe data yang berisi TRUE atau FALSE
* **str** : tipe data yang berisi text

### Tipe Data Categorical dan Numerical 

Ada beberapa tipe data yang cukup spesial seperti tipe data **categoric**. Tipe data categoric merupakan tipe data yang mirip dengan tipe data object, perbedaannya yaitu pada memori yang disimpan. Jika pada tipe data object maka untuk setiap karakter yang ada akan tersimpan pada suatu memori dengan id yang berbeda dengan karakter yang lain. Hal ini akan menyebabkan memori menyimpan terlalu banyak informasi terlebih apabila karakter yang ada pada data yang kita miliki terdapat pengulangan. Tentunya akan sangat memakan memori yang banyak. Oleh karena itu ada tipe data spesial yaitu **categoric** yang dapat kita gunakan untuk menyimpan memori data dengan lebih efisien. Cara kerja dari tipe data categoric yaitu akan menyimpan informasi dari suatu karakter yang unik dan akan disimpan dengan id tertentu sesuai dengan berapa banyak karakter unik yang muncul. 

Dalam beberapa kondisi, tipe data pada suatu kolom atau variabel sangatlah penting untuk diperhatikan. Hal ini akan mempermudah seorang data analyst mengetahui data apa yang sedang dia kerjakan. Selain itu juga ketika akan membuat suatu report, variabel tertentu dapat diinterpretasikan dengan baik menggunakan pengukuran statistik ataupun berupa visualisasi. Oleh karena itu, ketika tipe data dalam data frame ada yang tidak sesuai, kita perlu menyesuaikan agar data dapat direpresentasikan dengan baik. 

Pada course material, telah dibahas bagaimana mengubah tipe data pada suatu data frame. Untuk melakukan perubahan tipe data, kita bisa menggunakan fungsi `astype()` pada kolom yang ingin kita ubah.

**Knowledge check:**

Sebelum kita akan mengalisis kolom manakah yang belum sesuai, kita cek terlebih dahulu tipe data dari masing-masing kolom pada object `turnover` yang kita miliki.

In [18]:
## your code here
## Melihat tipe data

turnover.dtypes

satisfaction_level       float64
last_evaluation          float64
number_project             int64
average_montly_hours       int64
time_spend_company         int64
Work_accident              int64
promotion_last_5years      int64
division                  object
salary                    object
left                       int64
dtype: object

**Latihan!**

Jika kita perhatikan berdasarkan gloosary dari masing-masing kolom pada data `turnover` dan tipe datanya, bisakah Anda temukan kolom mana saja yang harus diubah dalam bentuk `category`? 

Jika sudah menemukannya, Anda dapat mengubahnya menggunakan fungsi `astype('category')` dan jangan lupa simpan kembali agar data `turnover` sudah memiliki tipe data yang tepat untuk masing-masing kolomnya.

In [19]:
## your code here
## Merubah beberapa tipe data pada kolom-kolom tertentu menjadi tipe category
turnover[["time_spend_company", "Work_accident", "promotion_last_5years", "division", "salary", "left"]] =\
turnover[["time_spend_company", "Work_accident", "promotion_last_5years", "division", "salary", "left"]].astype("category")

Lakukan pengecekan ulang mengenai tipe data yang sudah diubah, apakah sudah terubah dengan baik dengan menggunakan `dtypes` atribut.

In [20]:
## your code here
## Cek Ulang

turnover.dtypes


satisfaction_level        float64
last_evaluation           float64
number_project              int64
average_montly_hours        int64
time_spend_company       category
Work_accident            category
promotion_last_5years    category
division                 category
salary                   category
left                     category
dtype: object

### Quiz 2

Jika diketahui data sebagai berikut.

In [21]:
import pandas as pd

In [22]:
toko = pd.DataFrame({
    'jenis' : ['Sepatu', 'Sandal', 'Baju', 'Tas'],
    'warna' : ['Hitam', 'Merah', 'Biru', 'Pink'],
    'harga' : [1300000, 56000, 85000, 100000],
    'diskon' : [0.2, 0, 0, 0.1],
    'stok' : [59, 100, 200, 50],
    'penjualan_mingguan' : [6, 50, 150, 40]
})
toko

jenis  warna    harga  diskon  stok  penjualan_mingguan
0  Sepatu  Hitam  1300000     0.2    59                   6
1  Sandal  Merah    56000     0.0   100                  50
2    Baju   Biru    85000     0.0   200                 150
3     Tas   Pink   100000     0.1    50                  40

1. Jika Fafilia adalah seorang data analyst pada toko tersebuut, dia diminta untuk melakukan analisis terhadap data toko tersebut. Manakah dari kolom yang tersedia termasuk `category` atau `object`?
    - [ ] jenis dan harga
    - [ ] warna dan stok
    - [v] jenis dan warna
    - [ ] warna dan diskon
  
---

2. Manakah kolom yang termasuk `integer`?
    - [ ] diskon
    - [v] stok
    - [ ] jenis
    - [ ] warna

---

3. Manakah yang termasuk tipe data `float` pada data toko diatas?
    - [v] diskon
    - [ ] stok
    - [ ] jenis
    - [ ] penjualan_mingguan

## Eksplorasi Data

Selain mengetahui tipe data, terkadang kita juga perlu melakukan eksplorasi lebih lanjut mengenai data yang kita gunakan. Beberapa hal yang bisa kita lakukan dalam melakukan eksplorasi data adalah sebagai berikut :

- `.head()` dan `.tail()` : untuk menampilkan 5 data paling awal dan 5 data paling akhir
- `.dtypes` : untuk mengecek tipe data
- `.shape` : untuk mengetahui ukuran data tabel berdasarkan informasi baris dan kolomnya
- `.size` : untuk mengetahui besarnya data tabel berdasarkan informasi baris dikali kolom
- `.axes` : untuk mengetahui infomasi index names atau column names
- `.describe()` : untuk mengetahui summary singkat dari data 

Untuk melakukan inspek kolom yang kita perlukan, kita bisa menggunakan `df['namakolom']` jika hanya satu kolom atau dengan menggunakan `df[['kolom1','kolom2',...]]`. Misalkan kita ingin menampilkan 5 data pertama dari kolom `division`, `salary`, dan `left`.

In [23]:
turnover = pd.read_csv(r"C:\Users\Lenovo\Desktop\Machine Learning\data\turnover_balance.csv")
turnover.head()

## Modif lokasi penyimpanan datanya

satisfaction_level  last_evaluation  number_project  average_montly_hours  \
0                0.82             0.68               3                   140   
1                0.79             0.67               5                   156   
2                0.73             0.95               3                   149   
3                0.92             0.78               3                   218   
4                0.69             1.00               5                   237   

   time_spend_company  Work_accident  promotion_last_5years     division  \
0                   2              0                      0        sales   
1                   2              0                      0  product_mng   
2                   2              0                      0      support   
3                   3              0                      0    technical   
4                   3              0                      0    technical   

  salary  left  
0    low     0  
1    low     0  
2    low     0  
3    low     0  
4   high     0

In [24]:
turnover[['division', 'salary', 'left']].head()

division salary  left
0        sales    low     0
1  product_mng    low     0
2      support    low     0
3    technical    low     0
4    technical   high     0

Kita bisa lakukan pengecekan berapa banyak data yang tersedia pada `turnover` data dengan menggunakan `.shape` atau `.size`.

In [25]:
turnover.shape

(7142, 10)

In [26]:
turnover.size

71420

Untuk bisa melakukan inspek nama kolom kita bisa menggunakan atribut `.axes[1]` sedangkan untuk melakukan inspek nama index bisa menggunakan atribut `.axes[0]`.

In [27]:
turnover.axes[1]

Index(['satisfaction_level', 'last_evaluation', 'number_project',
       'average_montly_hours', 'time_spend_company', 'Work_accident',
       'promotion_last_5years', 'division', 'salary', 'left'],
      dtype='object')

Kita juga bisa menghitung banyaknya kolom yang ada pada suatu dataframe dengan menggunakan method `len()` dengan diikuti `df.axes[]`. Misalnya kita ingin menghitung banyaknya kolom pada dataframe `turnover`.

In [28]:
len(turnover.axes[1])

10

Terkadang kita juga membutuhkan informasi singkat dari rangkuman dari data yang kita miliki, kita bisa menggunakan `.describe()` untuk mendapatkan informasi dari rangkuman data kita.

In [29]:
turnover.describe()

satisfaction_level  last_evaluation  number_project  \
count         7142.000000      7142.000000     7142.000000   
mean             0.552569         0.715819        3.818818   
std              0.265043         0.181120        1.458167   
min              0.090000         0.360000        2.000000   
25%              0.380000         0.540000        2.000000   
50%              0.560000         0.730000        4.000000   
75%              0.780000         0.880000        5.000000   
max              1.000000         1.000000        7.000000   

       average_montly_hours  time_spend_company  Work_accident  \
count           7142.000000         7142.000000    7142.000000   
mean             203.366144            3.616074       0.108793   
std               54.195166            1.318875       0.311401   
min               97.000000            2.000000       0.000000   
25%              151.000000            3.000000       0.000000   
50%              205.000000            3.000000       0.000000   
75%              252.000000            4.000000       0.000000   
max              310.000000           10.000000       1.000000   

       promotion_last_5years         left  
count            7142.000000  7142.000000  
mean                0.016802     0.500000  
std                 0.128538     0.500035  
min                 0.000000     0.000000  
25%                 0.000000     0.000000  
50%                 0.000000     0.500000  
75%                 0.000000     1.000000  
max                 1.000000     1.000000

Informasi yang dikeluarkan oleh `.describe()` adalah 8 pengukuran deskriptif statistik berupa:
- Count
- Mean
- Standard Deviation
- Minimum Value
- 25th Percentile (Q1)
- 50th Percentile (Q2/Median)
- 75th Percentile (Q3)
- Maximum Value

Secara default, summary yang dikeluarkan oleh `.describe()` adalah summary dari kolom yang memiliki tipe data numeric. Apabila kita ingin menginputkan tipe data tertentu pada saat melakukan describe maka kita bisa gunakan parameter `include` untuk mendefinisikan tipe data yang akan dideskripsikan. 

Sebagai contoh disini kita akan coba menggunakan describe untuk tipe data object dan numeric.

In [30]:
turnover.describe(include = ['object', 'int64', 'float64'])

satisfaction_level  last_evaluation  number_project  \
count          7142.000000      7142.000000     7142.000000   
unique                 NaN              NaN             NaN   
top                    NaN              NaN             NaN   
freq                   NaN              NaN             NaN   
mean              0.552569         0.715819        3.818818   
std               0.265043         0.181120        1.458167   
min               0.090000         0.360000        2.000000   
25%               0.380000         0.540000        2.000000   
50%               0.560000         0.730000        4.000000   
75%               0.780000         0.880000        5.000000   
max               1.000000         1.000000        7.000000   

        average_montly_hours  time_spend_company  Work_accident  \
count            7142.000000         7142.000000    7142.000000   
unique                   NaN                 NaN            NaN   
top                      NaN                 NaN            NaN   
freq                     NaN                 NaN            NaN   
mean              203.366144            3.616074       0.108793   
std                54.195166            1.318875       0.311401   
min                97.000000            2.000000       0.000000   
25%               151.000000            3.000000       0.000000   
50%               205.000000            3.000000       0.000000   
75%               252.000000            4.000000       0.000000   
max               310.000000           10.000000       1.000000   

        promotion_last_5years division salary         left  
count             7142.000000     7142   7142  7142.000000  
unique                    NaN       10      3          NaN  
top                       NaN    sales    low          NaN  
freq                      NaN     1997   3773          NaN  
mean                 0.016802      NaN    NaN     0.500000  
std                  0.128538      NaN    NaN     0.500035  
min                  0.000000      NaN    NaN     0.000000  
25%                  0.000000      NaN    NaN     0.000000  
50%                  0.000000      NaN    NaN     0.500000  
75%                  0.000000      NaN    NaN     1.000000  
max                  1.000000      NaN    NaN     1.000000

Selain menggunakan `.describe()`, kita juga bisa menggunakan fungsi deksriptif statistik lain jika kita hanya ingin menampilkan spesifik nilai. Beberapa fungsi deskriptif statistik lain yang bisa digunakan yaitu:
- `.mean()` : menampilkan rata-rata
- `.sum()` : mendapatkan total nilai
- `.min()` dan `.max()` : mendapatkan nilai terkecil dan terbesar
- `.median()` : mendapatkan nilai tengah
- `.quantile()` : mendapatkan nilai quartile dari data, secara default akan memunculkan nilai median


> **!!!** Sedikit tips, kita bisa gunakan method `.info()` untuk menampilkan struktur suatu dataframe baik berupa banyaknya baris dan kolom, tipe data untuk masing-masing kolom, dan besarngya data yang kita gunakan.

In [31]:
turnover.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7142 entries, 0 to 7141
Data columns (total 10 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   satisfaction_level     7142 non-null   float64
 1   last_evaluation        7142 non-null   float64
 2   number_project         7142 non-null   int64  
 3   average_montly_hours   7142 non-null   int64  
 4   time_spend_company     7142 non-null   int64  
 5   Work_accident          7142 non-null   int64  
 6   promotion_last_5years  7142 non-null   int64  
 7   division               7142 non-null   object 
 8   salary                 7142 non-null   object 
 9   left                   7142 non-null   int64  
dtypes: float64(2), int64(6), object(2)
memory usage: 558.1+ KB


Sebagai data analyst, berdasarkan data yang kita miliki, kita ingin mencoba mencari tahu berapa besar rata-rata jam kerja bulanan dan rata-rata lama bekerja di perusahaan. Untuk mengetahui nilai rata-rata pada suatu series bisa menggunakan `.mean()`.

In [32]:
turnover["average_montly_hours"].mean()

203.36614393727248

Berdasarkan data yang ada, diketahui bahwa rata-rata pekerja yang ada yaitu sebesar 203 jam per bulan. Atau jika dikonversikan ke dalam hari yaitu rata-rata jam kerjanya sebesar 10 jam per hari.

In [33]:
turnover["time_spend_company"].mean()

3.6160739288714647

Berdasarkan data yang ada, diketahui bahwa rata-rata lama bekerja dalam suatu perusahaan yaitu 3.6 tahun. 

Sebelumnya, kita sudah memiliki data `turnover` dengan tipe data untuk masing-maisng kolom sebagai berikut.

In [34]:
turnover.dtypes

satisfaction_level       float64
last_evaluation          float64
number_project             int64
average_montly_hours       int64
time_spend_company         int64
Work_accident              int64
promotion_last_5years      int64
division                  object
salary                    object
left                       int64
dtype: object

Pada part sebelumnya kita sudah mencoba untuk mengubah kolom `division` dan `salary` tipe data object menjadi `category` dengan menggunakan method `astype()`. Jika kita perhatikan sekali lagi pada data `turnover`, masih terdapat beberapa kolom yang belum tepat tipe datanya berdasarkan informasi pada kolom tersebut. Oleh karena itu, coba ubah beberapa kolom yang seharusnya `category` dari kolom yang ada.

In [35]:
turnover.head()

satisfaction_level  last_evaluation  number_project  average_montly_hours  \
0                0.82             0.68               3                   140   
1                0.79             0.67               5                   156   
2                0.73             0.95               3                   149   
3                0.92             0.78               3                   218   
4                0.69             1.00               5                   237   

   time_spend_company  Work_accident  promotion_last_5years     division  \
0                   2              0                      0        sales   
1                   2              0                      0  product_mng   
2                   2              0                      0      support   
3                   3              0                      0    technical   
4                   3              0                      0    technical   

  salary  left  
0    low     0  
1    low     0  
2    low     0  
3    low     0  
4   high     0

In [36]:
## your code here

## Merubah beberapa tipe kolom menjadi tipe category

turnover[["time_spend_company", "Work_accident", "promotion_last_5years", "division", "salary", "left"]] =\
turnover[["time_spend_company", "Work_accident", "promotion_last_5years", "division", "salary", "left"]].astype("category")

turnover.dtypes

satisfaction_level        float64
last_evaluation           float64
number_project              int64
average_montly_hours        int64
time_spend_company       category
Work_accident            category
promotion_last_5years    category
division                 category
salary                   category
left                     category
dtype: object

### Quiz 3

Berdasarkan data `turnover` yang kita miliki, jawablah beberapa pertanyaan berikut ini.

1. Method apakah yang bisa kita gunakan untuk mendapatkan informasi tipe data, banyaknya kolom dan baris, serta nama-nama kolom dari dataframe `turnover`?
    - [ ] `.describe()`
    - [ ] `.head()`
    - [v] `.info()`
    - [ ] `.shape`
    
> **!!!** Sedikit tips, kita bisa gunakan method `.info()` untuk menampilkan struktur suatu dataframe baik berupa banyaknya baris dan kolom, tipe data untuk masing-masing kolom, dan besarngya data yang kita gunakan.

---

2. Berapakah rata-rata `satisfaction_level` yang diberikan oleh seluruh karyawan pada data `turnover`?
    - [v] 0.55
    - [ ] 0.56
    - [ ] 1.0
    - [ ] 0.78

> **!!!** turnover["satisfaction_level"].mean()
---

3. Jika kita ingin mengetahui nilai quartile ke-3 dari kolom yang memiliki tipe data numeric, manakah code dibawah ini yang tepat?
    - [ ] `turnover.quantile(0.25)`
    - [v] `turnover.quantile(0.75)`
    - [ ] `turnover.quantile(0.75, numeric_only=True)`
    
> **!!!** Selain menggunakan `.describe()`, kita juga bisa menggunakan fungsi deksriptif statistik lain jika kita hanya ingin menampilkan spesifik nilai. Beberapa fungsi deskriptif statistik lain yang bisa digunakan yaitu:
- `.quantile()` : mendapatkan nilai quartile dari data, secara default akan memunculkan nilai median
- 75th Percentile (Q3)

## Indexing dan Subsetting  

Jika berbicara mengenai analisis data, terkadang kita juga melakukan slicing atau subsetting pada data-data dengan kondisi tertentu. Secara umum apabila kita ingin melakukan subsetting dan indexing pada dataframe pada python yang disupport oleh `pandas`, kita bisa menggunakan beberapa cara berikut:

- `df[start:end]` : untuk melakukan subset berdasarkan index baris tertentu secara terurut
- `.loc[]` : untuk melakukan subset pada nama baris atau nama kolom tertentu
- `.iloc[]` : untuk melakukan subset menggunakan index baris atau kolom tertentu
- `df['column_name']` : untuk melakukan subset berdasarkan nama kolom tertentu
- `select_dtypes()` : untuk melakukan subset berdasarkan tipe data tertentu
- `.drop()` : untuk menghilangkan nilai tertentu

Beberapa cara diatas dapat digunakan sesuai dengan kondisi-kondisi tertentu. Beberapa sudah banyak dijelaskan pada course material **Python for Data Analyst**. 

Sebagai contoh misalnya kita hanya tertarik pada tipe data object pada data `turnover` untuk kita buat summary nya. Kita bisa menggunakan `select_dtypes()` untuk memilih tipe data yang kita inginkan.

In [37]:
turnover = pd.read_csv(r"C:\Users\Lenovo\Desktop\Machine Learning\data\turnover_balance.csv")

In [38]:
turnover.select_dtypes('object').describe()    # select_dtypes("object") membaca tipe data object saja

division salary
count      7142   7142
unique       10      3
top       sales    low
freq       1997   3773

Misalnya kita ingin melakukan subset data ke-8 hingga ke-15 pada kolom 'average_montly_hours' dan 'time_spend_company' dimana jika dalam index pada data karena python memiliki index yang berawal dari angka 0 maka kita ingin mengambil data index ke-7 hingga ke-14. 

Urutan nilai pada python akan diawali dengan index ke 0 hingga index-1, jadi apabila kita menuliskan urutan index ke-7 hingga ke-14, maka yang akan muncul adalah data dari index ke-7 hingga ke-13.

Oleh karena itu apabila kita ingin melakukan subset untuk data ke-8 hingga ke-15 maka kita tulis indexnya dari 7 hingga 15.

In [39]:
turnover[7:15]

satisfaction_level  last_evaluation  number_project  average_montly_hours  \
7                 0.51             0.73               4                   229   
8                 0.88             0.99               3                   190   
9                 0.76             0.85               3                   192   
10                0.55             0.55               4                   257   
11                0.43             0.46               2                   186   
12                0.80             0.87               5                   251   
13                0.64             0.67               5                   167   
14                0.84             0.37               5                   186   

    time_spend_company  Work_accident  promotion_last_5years     division  \
7                    3              0                      0        sales   
8                    5              0                      0    technical   
9                    3              0                      0    marketing   
10                   3              0                      0  product_mng   
11                   2              0                      0  product_mng   
12                   3              0                      0        sales   
13                   2              0                      0        sales   
14                   2              0                      0    technical   

    salary  left  
7      low     0  
8     high     0  
9      low     0  
10     low     0  
11     low     0  
12  medium     0  
13     low     0  
14  medium     0

In [40]:
turnover[7:15].loc[:,['average_montly_hours','time_spend_company']]

average_montly_hours  time_spend_company
7                    229                   3
8                    190                   5
9                    192                   3
10                   257                   3
11                   186                   2
12                   251                   3
13                   167                   2
14                   186                   2

Penggunaan code diatas akan sama dengan code dibawah ini jika kita menerapkan penggunaan `loc[]` dan `iloc[]`.

In [41]:
turnover.iloc[7:16,:].loc[:,['average_montly_hours','time_spend_company']]

average_montly_hours  time_spend_company
7                    229                   3
8                    190                   5
9                    192                   3
10                   257                   3
11                   186                   2
12                   251                   3
13                   167                   2
14                   186                   2
15                   224                   2

Kebalikan dari `loc[]` dan `iloc[]`, `drop()` dapat kita gunakan untuk menghapus nilai tertentu pada data. 

Misalnya kita ingin menghapus kolom 'average_montly_hours' dan 'time_spend_company' pada data `turnover`.

In [42]:
turnover.drop([ 'average_montly_hours','time_spend_company'], axis=1)

satisfaction_level  last_evaluation  number_project  Work_accident  \
0                   0.82             0.68               3              0   
1                   0.79             0.67               5              0   
2                   0.73             0.95               3              0   
3                   0.92             0.78               3              0   
4                   0.69             1.00               5              0   
...                  ...              ...             ...            ...   
7137                0.11             0.89               6              0   
7138                0.77             0.87               4              0   
7139                0.79             1.00               4              1   
7140                0.88             0.97               4              1   
7141                0.37             0.47               2              0   

      promotion_last_5years     division  salary  left  
0                         0        sales     low     0  
1                         0  product_mng     low     0  
2                         0      support     low     0  
3                         0    technical     low     0  
4                         0    technical    high     0  
...                     ...          ...     ...   ...  
7137                      0        sales  medium     1  
7138                      0        sales     low     1  
7139                      0   management  medium     1  
7140                      0           IT     low     1  
7141                      0           hr     low     1  

[7142 rows x 8 columns]

Tujuan penggunaan parameter `axis = 1` adalah agar dalam memberikan perintah `drop()` langsung tertuju pada kolom. 
Jadi apabila kita ingin menghapus baris tertentu kita bisa gunakan `axis = 0`, dimana secara default method `drop()` menggunakan `axis = 0`.

In [43]:
turnover.drop([7,8,9,10,11,12,13,14,15]).head(15)

satisfaction_level  last_evaluation  number_project  average_montly_hours  \
0                 0.82             0.68               3                   140   
1                 0.79             0.67               5                   156   
2                 0.73             0.95               3                   149   
3                 0.92             0.78               3                   218   
4                 0.69             1.00               5                   237   
5                 0.98             0.97               3                   209   
6                 0.52             0.90               4                   285   
16                0.68             0.97               3                   250   
17                0.47             0.46               2                   152   
18                0.80             0.62               3                   191   
19                0.58             0.43               6                   132   
20                0.74             0.85               5                   135   
21                0.81             0.62               4                   153   
22                0.89             0.71               4                   190   
23                0.27             0.76               4                   108   

    time_spend_company  Work_accident  promotion_last_5years     division  \
0                    2              0                      0        sales   
1                    2              0                      0  product_mng   
2                    2              0                      0      support   
3                    3              0                      0    technical   
4                    3              0                      0    technical   
5                    3              0                      0      support   
6                    2              0                      0           hr   
16                   3              1                      0  product_mng   
17                   2              0                      0           IT   
18                   3              0                      0      support   
19                   3              0                      0        sales   
20                   2              0                      0      support   
21                   4              0                      0        RandD   
22                   3              0                      0        RandD   
23                  10              0                      0        sales   

    salary  left  
0      low     0  
1      low     0  
2      low     0  
3      low     0  
4     high     0  
5      low     0  
6      low     0  
16  medium     0  
17  medium     0  
18  medium     0  
19     low     0  
20     low     0  
21     low     0  
22     low     0  
23  medium     0

### Kondisional Subset 

Selain melakukan subsetting berdasarkan nama baris dan kolomnya, kita juga bisa melakukan subsetting berdasarkan kondisi tertentu. Untuk melakukan subsetting dengan kondisi tertentu ada beberapa **logical operator** yang bisa kita gunakan yaitu sebagai berikut.

- `==` sama dengan
- `!=` tidak sama dengan
- `>` lebih besar
- `<` lebih kecil
- `>=` lebih besar sama dengan
- `<=` lebih kecil sama dengan
- `&` AND
- `|` OR


Misalkan disini kita ingin mencari tahu berapa banyak karyawan yang berada pada divisi sales.

In [44]:
turnover[turnover['division'] == 'sales']

satisfaction_level  last_evaluation  number_project  \
0                   0.82             0.68               3   
7                   0.51             0.73               4   
12                  0.80             0.87               5   
13                  0.64             0.67               5   
19                  0.58             0.43               6   
...                  ...              ...             ...   
7128                0.74             1.00               5   
7133                0.09             0.77               6   
7134                0.39             0.57               2   
7137                0.11             0.89               6   
7138                0.77             0.87               4   

      average_montly_hours  time_spend_company  Work_accident  \
0                      140                   2              0   
7                      229                   3              0   
12                     251                   3              0   
13                     167                   2              0   
19                     132                   3              0   
...                    ...                 ...            ...   
7128                   275                   5              0   
7133                   282                   5              0   
7134                   127                   3              0   
7137                   301                   4              0   
7138                   242                   6              0   

      promotion_last_5years division  salary  left  
0                         0    sales     low     0  
7                         0    sales     low     0  
12                        0    sales  medium     0  
13                        0    sales     low     0  
19                        0    sales     low     0  
...                     ...      ...     ...   ...  
7128                      0    sales     low     1  
7133                      0    sales  medium     1  
7134                      0    sales     low     1  
7137                      0    sales  medium     1  
7138                      0    sales     low     1  

[1997 rows x 10 columns]

Jika kondisinya lebih dari satu, maka kita harus memisahkan setiap kondisi dengan tanda kurung `()`. Misal kita ingin tahu yang termasuk divisi sales dan time_spend_company nya lebih dari 3 tahun.

In [45]:
turnover[['division', 'salary']] = turnover[['division', 'salary']].astype('category')
turnover['division'].cat.categories

Index(['IT', 'RandD', 'accounting', 'hr', 'management', 'marketing',
       'product_mng', 'sales', 'support', 'technical'],
      dtype='object')

### Quiz 4

Berdasarkan data `turnover` coba lakukan subsetting dan jawablah pertanyaan dibawah ini.

1. Berapa banyak karyawan  yang memberikan penilaian satisfaction_level lebih dari 0.8?
2. Katakanlah kita sedang di posisi HR yang ingin melakukan analisis terhadap karyawan yang memberikan satisfaction_level kurang dari 0.5, tingkat salary yang rendah dan memiliki status left 0. Berapa banyak karyawan yang harus diperhatikan oleh HR?
3. Sebagai seorang HR, kita ingin coba menerapkan beberapa treatment khusus bagi karyawan yang berada pada divisi HR  yang tidak mendapatkan promosi selama 5 tahun belakang (`promotion_last_5years`) dan memiliki tingka salary yang rendah. Berapa banyak karyawan yang perlu perhatian lebih oleh HR?

#### Jawaban Quiz 4

##### 1.  Berapa banyak karyawan yang memberikan penilaian satisfaction_level lebih dari 0.8?

In [46]:
## Alternatif 1 Kalau mau ditampilkan datanya

turnover[turnover["satisfaction_level"] >= 0.8].loc[:, ["satisfaction_level"]]

# ada 1678 Row data

satisfaction_level
0                   0.82
3                   0.92
5                   0.98
8                   0.88
12                  0.80
...                  ...
7123                0.80
7124                0.86
7127                0.81
7129                0.82
7140                0.88

[1678 rows x 1 columns]

In [47]:
## Alternatif 2 Kalau mau Hasil angkanya saja

turnover[turnover["satisfaction_level"] >= 0.8].loc[:, ["satisfaction_level"]].count()


satisfaction_level    1678
dtype: int64

##### 2.  Katakanlah kita sedang di posisi HR yang ingin melakukan analisis terhadap karyawan yang memberikan satisfaction_level kurang dari 0.5, tingkat salary yang rendah dan memiliki status left 0. Berapa banyak karyawan yang harus diperhatikan oleh HR?

In [48]:
## Alternatif 1 kalau mau melihat datanya

turnover[
    (turnover["satisfaction_level"] < 0.5 ) &
    (turnover["salary"] == "low" ) &
    (turnover["left"] == 0 )
].loc[:,["satisfaction_level", "salary", "left"]]

satisfaction_level salary  left
11                  0.43    low     0
24                  0.38    low     0
25                  0.44    low     0
31                  0.49    low     0
37                  0.28    low     0
...                  ...    ...   ...
3505                0.41    low     0
3526                0.48    low     0
3534                0.35    low     0
3542                0.26    low     0
3543                0.31    low     0

[277 rows x 3 columns]

In [49]:
## Alternatif 2 kalau mau melihat hasilnya saja/angka hasilnya

turnover[
    (turnover["satisfaction_level"] < 0.5 ) &
    (turnover["salary"] == "low" ) &
    (turnover["left"] == 0 )
].loc[:,["satisfaction_level", "salary", "left"]].count()

satisfaction_level    277
salary                277
left                  277
dtype: int64

##### 3.  Sebagai seorang HR, kita ingin coba menerapkan beberapa treatment khusus bagi karyawan yang berada pada divisi HR yang tidak mendapatkan promosi selama 5 tahun belakang (promotion_last_5years) dan memiliki tingka salary yang rendah. Berapa banyak karyawan yang perlu perhatian lebih oleh HR?

In [50]:
## Alternatif 1 kalau mau melihat datanya

turnover[
    (turnover["promotion_last_5years"] == 0 ) &
    (turnover["salary"] == "low" )
].loc[:,["promotion_last_5years", "salary"]]

## Hasilnya 3741 / banyak juga yang gak naik-naik pangkatnya

promotion_last_5years salary
0                         0    low
1                         0    low
2                         0    low
3                         0    low
5                         0    low
...                     ...    ...
7135                      0    low
7136                      0    low
7138                      0    low
7140                      0    low
7141                      0    low

[3741 rows x 2 columns]

In [51]:
## Alternatif 2 kalau mau melihat hasilnya / angka

turnover[
    (turnover["promotion_last_5years"] == 0 ) &
    (turnover["salary"] == "low" )
].loc[:,["promotion_last_5years", "salary"]].count()


promotion_last_5years    3741
salary                   3741
dtype: int64

# Task

Selamat!! Anda sudah selesai mempelajari beberapa pemahaman dasar dalam melakukan analisis data menggunakan `pandas`. Sebagai bentuk Anda dalam mempelajari pemahaman dasar menggunakan `pandas`, kita akan coba berikan beberapa soal untuk menguji pemahaman Anda mengenai beberapa materi diatas. 

Pada folder data_input terdapat file .csv dengan nama handphone.csv. Lakukan import data `handphone.csv` menggunakan `pd.read_csv()` dengan menambahkan parameter `encoding='unicode_escape'`. Setelah berhasil terbaca datanya, kerjakanlah beberapa soal dibawah ini. 

Data `handphone.csv` merupakan data penjualan handphone dari beberapa seller yang dipromosikan di suatu e-commerce di Indonesia. Berikut ini beberapa penjelasan untuk masing-masing kolomnya.
- `product_title` : nama produk yang di promosikan di laman e-commerce
- `location` : lokasi seller
- `account_seller` : nama akun seller
- `price` : harga produk
- `review` : total review yang diberikan untuk produk yang dijual seller
- `installment` : besarnya cicilan yang harus dibayarkan (jika menggunakan metode cicilan untuk membayar)
- `rating` : rating untuk produk yang dijual oleh seller
- `brand_hp` : brand handphone yang dijual

1. Adakah tipe data yang belum sesuai dari kolom pada data tersebut? Jika ada, ubahlah kedalam tipe data yang sesuai.

In [52]:
## your code here

handphone = pd.read_csv(r"C:\Users\Lenovo\Desktop\Machine Learning\data\handphone.csv", encoding = "unicode_escape")

handphone

product_title        location  \
0                                 Ken Mobile R7710 Grey         Jakarta   
1     OPPO A3s Smartphone AI Beauty 2.0 Camera 2GB/1...       Tangerang   
2        Asus Zenfone 5 ZE620KL 4/64 GARANSI RESMI ASUS         Jakarta   
3     OPPO A3s Smartphone AI Beauty 2.0 Camera 2GB/1...       Tangerang   
4                   Sony Xperia Z4 Compact Docomo Seken  Kab. Tangerang   
...                                                 ...             ...   
4915           realme2 pro 8/128gb Garansi Resmi 1tahun           Jambi   
4916  NOKIA 130 NEW 2017 GARANSI RESMI 1 THN, ORIGIN...         Jakarta   
4917  XIAOMI REDMI NOTE 6 PRO 4/64GB RAM 4GB ROM 64G...         Jakarta   
4918  Ori HP VIVO V7  / VIVO V7 PLUS-RAM 4/64GB - BL...     Banjarmasin   
4919                       Sony Xperia Z3 Big ex Docomo         Jakarta   

           account_seller      price  review  installment  rating brand_hp  
0              KEN Mobile   199000.0     432       8292.0     4.7      ken  
1     OPPO OFFICIAL STORE  1699000.0     144      70792.0     4.8     oppo  
2            Cellular Mas  2830000.0     454     117917.0     4.9     asus  
3     OPPO OFFICIAL STORE  1699000.0     195      70792.0     4.8     oppo  
4      Five star cellular   530000.0     711      22084.0     4.6     sony  
...                   ...        ...     ...          ...     ...      ...  
4915        younglic shop  4250000.0      37     177084.0     5.0   realme  
4916       GADING BINTANG   285000.0     123      11875.0     4.9    nokia  
4917     Kreasindo Online  2390000.0     123      99584.0     4.9   xiaomi  
4918             REDMARTT  4222000.0       9     175917.0     4.9     vivo  
4919             tokoChoy  1500000.0      45      62500.0     5.0     sony  

[4920 rows x 8 columns]

In [53]:
handphone.dtypes

product_title      object
location           object
account_seller     object
price             float64
review              int64
installment       float64
rating            float64
brand_hp           object
dtype: object

In [54]:
handphone[["account_seller", "brand_hp", "location"]] = \
handphone[["account_seller", "brand_hp", "location"]].astype("category")

In [55]:
handphone.dtypes

product_title       object
location          category
account_seller    category
price              float64
review               int64
installment        float64
rating             float64
brand_hp          category
dtype: object

---

2. Dimanakah lokasi seller terbanyak yang menjual handphone?
    - [ ] Tangerang
    - [ ] Yogyakarta
    - [v] Jakarta
    - [ ] Bekasi

---

3. Berapa banyak seller yang mempromosikan handphone di e-commerce tersebut yang berlokasi di Jakarta?
    - [ ] 4920
    - [ ] 116
    - [ ] 156
    - [v] 2723

---

4. Berapakah rata-rata rating yang diberikan oleh pembeli untuk seller 'OPPO OFFICIAL STORE'?
    - [v] 4.88
    - [ ] 4.71
    - [ ] 4.59
    - [ ] 4.63

---

5. Berapakah rata-rata harga handphone dengan brand oppo yang dijual oleh 'OPPO OFFICIAL STORE'?
    - [ ] Rp 1.799.000
    - [v] Rp 3.141.307
    - [ ] Rp 4.049.450
    - [ ] Rp 1.913.800

---

6. Berapakah rating terendah yang diberikan oleh pembeli kepada 'OPPO OFFICIAL STORE'? 
    - [ ] 4.5
    - [ ] 3.5
    - [ ] 4.6
    - [v] 4.8
    
---

7. Berapakah produk yang memiliki rating kurang dari 4?
    - [ ] 18
    - [v] 14
    - [ ] 218
    - [ ] 15

---

8. Brand handphone apakah yang paling banyak dipromosikan melalui e-commerce tersebut?
    - [ ] oppo
    - [ ] apple
    - [v] xiaomi
    - [ ] asus

---

9. Seller manakah yang memberikan installment paling rendah dari produk yang dipasarkan? Silahkan pilih lebih dari 1 jawaban dibawah ini.
    - [v] Shopuz ID
    - [v] enterphone2
    - [ ] Kelontong Unik
    - [ ] ERAFONE

In [56]:
## Soal nomor 2 ##

handphone["location"].describe()

count        4920
unique         97
top       Jakarta
freq         2723
Name: location, dtype: object

In [57]:
## Soal nomor 3 ##

handphone[
    (handphone["location"] == "Jakarta")
]

product_title location  \
0                                 Ken Mobile R7710 Grey  Jakarta   
2        Asus Zenfone 5 ZE620KL 4/64 GARANSI RESMI ASUS  Jakarta   
5     Xiaomi Redmi 6A 2/16 RAM 2GB ROM 16GB GARANSI ...  Jakarta   
8            Samsung Galaxy J2 Prime GARANSI RESMI SEIN  Jakarta   
9     Samsung Galaxy A50 4/64 RAM 4GB ROM 64GB GARAN...  Jakarta   
...                                                 ...      ...   
4910       OPPO A37 ( ROSE GOLD ) NEW GRS RESMI 1 TAHUN  Jakarta   
4912  Samsung Galaxy A10 2/32 RAM 2GB ROM 32GB GARAN...  Jakarta   
4916  NOKIA 130 NEW 2017 GARANSI RESMI 1 THN, ORIGIN...  Jakarta   
4917  XIAOMI REDMI NOTE 6 PRO 4/64GB RAM 4GB ROM 64G...  Jakarta   
4919                       Sony Xperia Z3 Big ex Docomo  Jakarta   

              account_seller      price  review  installment  rating brand_hp  
0                 KEN Mobile   199000.0     432       8292.0     4.7      ken  
2               Cellular Mas  2830000.0     454     117917.0     4.9     asus  
5               Cellular Mas  1185000.0    1133      49375.0     4.9   xiaomi  
8               Cellular Mas  1125000.0    2486      46875.0     4.9  samsung  
9               Cellular Mas  3380000.0     388     140834.0     4.9  samsung  
...                      ...        ...     ...          ...     ...      ...  
4910         Grandstore cell  1699000.0     259      70792.0     4.7     oppo  
4912  JUAL HANDPHONE DOT COM  1699000.0       7      70792.0     5.0  samsung  
4916          GADING BINTANG   285000.0     123      11875.0     4.9    nokia  
4917        Kreasindo Online  2390000.0     123      99584.0     4.9   xiaomi  
4919                tokoChoy  1500000.0      45      62500.0     5.0     sony  

[2723 rows x 8 columns]

In [58]:
# Soal nomor 4, 5, 6 ##

handphone[
    (handphone["account_seller"] == "OPPO OFFICIAL STORE")
].describe()

price      review    installment     rating
count  1.300000e+01   13.000000      13.000000  13.000000
mean   3.141308e+06   69.769231  130888.076923   4.884615
std    1.108649e+06   62.057975   46193.634322   0.068874
min    1.699000e+06    9.000000   70792.000000   4.800000
25%    1.999000e+06   26.000000   83292.000000   4.800000
50%    3.149000e+06   47.000000  131209.000000   4.900000
75%    3.699000e+06  133.000000  154125.000000   4.900000
max    4.699000e+06  195.000000  195792.000000   5.000000

In [59]:
## Soal nomor 7 ##

handphone[
    (handphone["rating"] < 4)
].count()

product_title     14
location          14
account_seller    14
price             14
review            14
installment       14
rating            14
brand_hp          14
dtype: int64

In [60]:
## Soal nomor 8 ##

handphone["brand_hp"].value_counts()

xiaomi          1404
samsung         1095
apple            559
oppo             441
lain-lain        306
vivo             243
nokia            209
asus             178
realme           115
sony             107
huawei            47
lg                24
evercoss          22
sharp             19
lenovo            17
ken               16
meizu             16
blu               16
google pixel      14
motorola          11
polytron           9
smartfren          9
infinix            8
honor              7
cat                6
zte                4
oneplus            4
leeco              3
wiko               3
acer               3
htc                2
orico              1
philips            1
siemens            1
Name: brand_hp, dtype: int64

In [61]:
## Soal nomor 9 ##

handphone.sort_values(by = ["installment"])

product_title        location  \
484                     Buble Wrap For Safety Packaging         Jakarta   
3579  Adaptor Colokan Xiaomi pipih / gepeng (US) ke ...         Jakarta   
4342  Softcase Vivo V5 Plus X9 V5+ 5.5 inchi Ultrath...         Jakarta   
4156  PEGANGAN HP MOBIL UNIVERSAL STEERING WHEEL CAR...         Jakarta   
3622  Tempered Glass Huawei Y3-II Y3 2 Y3-11 Y3-2 4....         Jakarta   
...                                                 ...             ...   
3173  READY MAX XS 512GB APPLE IPHONE ORIGINAL MAX X...  Kab. Tangerang   
3171  APPLE IPHONE XSMAX 256GB DUAL SIM NANO+NANO GA...         Jakarta   
4196  (DUAL SIM) iPhone XS MAX 2018 256GB 256 Gold G...         Jakarta   
1946  ( DUAL SIM ) iPhone XS MAX 256GB 256 Gold Grey...         Jakarta   
2513                           Orderan Juragan Quantity         Jakarta   

           account_seller       price  review  installment  rating   brand_hp  
484           enterphone2      5000.0      87        209.0     4.8  lain-lain  
3579            Shopyz ID      5000.0      26        209.0     4.3     xiaomi  
4342    Agen Aksesoris Hp      5700.0      17        238.0     4.4       vivo  
4156          GBU JAKARTA      7000.0      18        292.0     4.1  lain-lain  
3622    Agen Aksesoris Hp      7700.0      29        321.0     4.3     huawei  
...                   ...         ...     ...          ...     ...        ...  
3173    online phone shop  18000000.0      26     750000.0     4.9      apple  
3171           Kios Utama  18599000.0      41     774959.0     5.0      apple  
4196          Putra Group  18850000.0      19     785417.0     5.0      apple  
1946    Iphone X Original  18850000.0      48     785417.0     5.0      apple  
2513  BZT Cellular Dan IT  20350000.0      57     847917.0     4.7  lain-lain  

[4920 rows x 8 columns]